src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -119,20 +119,22 @@ private static T[] LazyToArrayUsingBufferList<T>(IEnumerator<T> en, T[] first)
 
             // Here's a visualization showing what first, buffers, and current
             // might look like for a 200-length enumerable:
-            //
-            // first: [items 0-31]
-            //
-            // buffers:
-            // [0]: [items 32-63]
-            // [1]: [items 64-127]
-            //
-            // current: [items 128-199], [slots 200-255 empty]
-
-            // The up-front allocation for the list will not be that much since
-            // the backing store will be Array.Empty<T>() if you do not pass in a
-            // capacity. It only starts allocating arrays when we add the first item.
-
-            var buffers = new List<T[]>(); // list of previous buffers
+            
+            /*
+                first: [items 0-31]
+
+                buffers:
+                [0]: [items 32-63]
+                [1]: [items 64-127]
+
+                current: [items 128-199], [slots 200-255 empty]
+            */
+
+            // There will be no up-front allocation for the ArrayBuilder, since
+            // the backing store will be nil if you do not pass in a capacity.
+            // It only starts allocating arrays when we add the first item.
+
+            var buffers = new ArrayBuilder<T[]>(); // list of previous buffers
```

Comment:
Note: Since @jkotas suggested changing the resizing algorithm, the allocation pattern here will change slightly. Before we would allocate a `T[][4]` for 65-128 enumerables, and then the next allocation of a `T[][8]` would come for 1025+ enumerables. Now, we allocate a `T[][1]` for length 65, and then a `T[][3]` for length 129, and then a `T[][7]` for length 513.